### PR TITLE
Option B: keep functions unbound, replace `this` with `typed.referTo(...)` and `typed.referToThis(...)`

### PR DESCRIPTION
**Pull request by josdejong**
Keep typed-functions unbound, so they can be bound to a context by the user. Replaces `this(...)` with callbacks:

- `typed.referTo(signature1, signature2, ..., (fn1, fn2, ...) => { ... })` 
- `typed.referToSelf((self) => { ... })`

See discussion in #126 for pros/cons

To do:

- [x] Discuss API
- [x] Implement final API
- [x] Write tests
- [x] Write documentation 
- [x] Backward compatibility warnings. This PR is quite a breaking change, and I expect users will get very vague errors when using `this(...)` in a typed-function when upgrading from v2 to v3. Therefore I would like to scan the functions for occurrences of `this(...)` or `this.signatures`, and if so throw an exception. There must be an option to disable this validation step.
- [x] Refactor `resolveReferences`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -390,24 +390,45 @@ A typed function can be constructed in two ways:
...
-  'string': function (value) {
-    // on the following line we self reference the typed-function using "this"
-    return this(parseInt(value, 10));
-  }
-});
+    ```
+    typed.referTo(signature1, signature2, ..., function callback(fn1, fn2, ...) {
+      // ... use the resolved signatures fn1, fn2, ...
+    }) 
+    ```
```
Comment (gwhitney) [inline README.md:477]: There is some trailing whitespace on line 400 and several other places; I guess there is no linting implemented in this repo as of yet.

File: README.md [hunk 2]
```diff
@@ -390,24 +390,45 @@ A typed function can be constructed in two ways:
...
+        return 'Input was a number: ' + value;
+      },
+      'boolean': function (value) {
+        return 'Input was a boolean: ' + value;
+      },
+      'string': typed.referTo('number', 'boolean', (fnNumber, fnBoolean) => {
+        return function fnString(value) {
+          // here we use one of the other signatures of the typed-function directly:
+          return fnNumber(parseInt(value, 10));
+        }
```
Comment (gwhitney) [inline README.md:499]: I'd suggest this example would be more illustrative/"convincing" if both of the other signatures were actually used in the body, so I'd recommend adding
```
  if (value === 'true') return fnBoolean(true);
  if (value === 'false') return fnBoolean(false);
```
before the return fnNumber...

File: test/construction.test.js [hunk 3]
```diff
@@ -355,17 +355,146 @@ describe('construction', function() {
...
+          }
+        }),
+        // we define `number` after we use it in `string` to enforce getting an error
+        'number': typed.referTo(() => {
+          return 'number:' + value;
+        })
+      });
+    }, /Cannot refer to signature "string": signature is referring to a signature "number" which is not yet resolved/)
+  })
+
```
Comment (gwhitney) [inline test/construction.test.js:465]: There's no actual reason this example has to fail; typed-function could resolve number first, and then resolve string. And in fact, there is practical reason to support that style: in mathjs, it is often the case that some simple signature refers to a more elaborate one, and then than more elaborate one refers to the final, implemented one, and they are listed in that order, which is a very readable order for a human trying to understand what's going on. I wouldn't want this change to force the re-ordering of the presentation of those different implementations. So I'd suggest that the ideal would be for typed-function to resolve all the signatures it can, and then resolve more, etc., and only throw an error if it makes no progress on a pass and there are as-of-yet-unresolved signatures (which would imply that there was a reference loop in the signatures, so the error could say so). (Alternatively, you could sort the signatures by what refers to what first, but you'd have to make sure you used a sort algorithm that would terminate in the face of loops in the order, and then resolve the sorted signatures -- whichever seems easier.)

Of course this is just a suggestion, and you could leave things as they are where it's the burden of the user to order the signatures so there are no forward references, but in that case I would mention it in the documentation above. I think it would be nice, though, if the infrastructure took this burden off and allowed clients to present the signatures in the "clearest" order even if that means forward references.

Naturally, if you take this suggestion then this test should be split in two: one with a forward reference that works, and one with a reference loop that throws an error.

File: typed-function.js [hunk 4]
```diff
@@ -965,21 +965,102 @@
...
-    function createTypedFunction(name, signaturesMap) {
-      if (Object.keys(signaturesMap).length === 0) {
+    function createTypedFunction(name, rawSignaturesMap) {
+      if (Object.keys(rawSignaturesMap).length === 0) {
         throw new SyntaxError('No signatures provided');
       }
 
+      // resolve references, providing "self"
+      var signaturesMap = resolveReferences(rawSignaturesMap, fn);
+
```
Comment (gwhitney) [inline typed-function.js]: I would recommend that you resolve references _after_ splitting the signatures. Right now the following fails with `reference signature "string" not found`, even though the typed-function is perfectly well defined for strings:
```
const refer = typed({
  'number|string': arg => 'wow: ' + arg,
  'boolean': typed.referTo('string', refS => b => {
    if (b) return refS('true that!')
    return refS('no way...')
  })
})
```

File: typed-function.js [hunk 5]
```diff
@@ -965,21 +965,102 @@
...
+     * there is an issue with a resolved reference
+     * @param {function | undefined} resolvedReference
+     * @param {string} reference
+     * @param {string} signature
+     */
+    function verifyResolvedReference(resolvedReference, reference, signature) {
+      if (!resolvedReference) {
+        throw new TypeError('Cannot refer to signature "' + signature + '": ' +
+          'reference signature "' + reference + '" not found');
+      }
```
Comment (gwhitney) [inline typed-function.js:1037]: I think this error message should read 'Cannot resolve reference in signature "blah blah": reference signature "other" not found', since there is not necessarily any reference _to_ that first signature (the one in variable `signature` here).

File: typed-function.js (hunk 4 above)
Comment (gwhitney) [reply typed-function.js]: This is actually critical, so I think not just a recommendation, but a necessity. If you resolve references before splitting the signatures, then you would have to write the above example as
```
const refer = typed({
  'number|string': arg => 'wow: ' + arg,
  'boolean': typed.referTo('number|string', refS => b => {
    if (b) return refS('true that!')
    return refS('no way...')
  })
})
```
But now this function is unmergeable:
```
const reref = typed(refer, typed({Array: a => 'Got ' + a.length + 'items'}))
```
results in
```
TypeError: Cannot refer to signature "boolean": reference signature "number|string" not found
```
because of course once it's fully constructed, `refer` has no signature "number|string", it has two signatures "number" and "string" that correspond to the same implementation (`arg => 'wow' + arg`).
[By the way, it was trying to understand this example that led to my filing #141 and the associated PR #142.]

File: typed-function.js [hunk 6]
```diff
@@ -965,21 +965,102 @@
...
+
+              return resolvedReference;
+            })
+            var resolved = fn.referTo.callback.apply(this, resolvedReferences);
+
+            // we need to keep the .referTo object (with references and callback)
+            // to be able to merge typed-functions later on: then we need to
+            // resolve the new signatures again.
+            resolved.referTo = fn.referTo;
+
```
Comment (gwhitney) [inline typed-function.js:1103]: While intuitively the comment here before line 1025 is correct, if you comment out line 1025 all of the unit tests succeed. But that's no so surprising, because as far as I could see there are no unit tests for merging functions that have "refer-to"s in them. It seems to me that there certainly should be, so that's one recommended action item here.

But further, I was not actually able to create a test that failed even with this line commented out: there does not seem to be any way to override an existing signature of a typed function in a merge, and there is no way for a merge to "capture" a reference from one of the the merged items to another. That's because a single plain function can't be a refer-to, and so refer-to is only relevant to the typed-function arguments; but since all of the typed-function arguments had to have been fully resolved at their creation times, all of their references were satisfied "internally" to each of themselves, and so after the merge all of them will still resolve to exactly the same thing that they would have resolved to on their own.

But note this is **not** a recommendation to remove this code propagating the .referTo object. It **is** (partly) an argument to accept PR #142, because then one could merge new additional signatures with an existing typed-function that would referTo signatures that were on that existing typed-function, greatly increasing the power to leverage existing behavior when extending.

But even with #142, you're only resolving new "referTo"s, not re-resolving previously resolved "referTo"s  and arriving at a new outcome. That's because of the prohibition on overriding an existing signature. However, I _still_ think the referTo property should be kept, because I think that sooner or later typed-functions _should_ have the ability to modify a previously existing signature by some reasonable API, and for that purpose it would be best to have been keeping the referTo around.

 (But all that said if you wanted to drop the referTo for now, I racked my brains but could not come up with a way under the current API that it could make a difference, or even with #142 present make a difference. If you have such an example, it should _definitely_ be captured in a unit test, I should think.)

**Comment by gwhitney** [comment]
> * Backward compatibility warnings.

I thought about this a bit too and here's one suggestion: since formerly typed-function took over "this", typical calls "out there" will likely generally have the incoming "this" be null or undefined. So as an interim measure for compatibility errors, when incoming this is false-y, typed-function could replace it with a function that issues an error with a clear message about the change. (The only other ways I could think of to scan involved calling the implementations, but I don't see any way to avoid potential side effects if you do.) Anyhow, just a thought in case it helps.

File: README.md (hunk 1 above)
Comment (josdejong) [reply README.md:477]: Ah, indeed. Yeah there is no linting in place, I miss that 😅 

I'll add a semicolon there `;` to keep the code styling consistent, using `;` everywhere. (would love to change that though, the library is still using `;`, `var`, UMD, quite previous generation)

File: README.md (hunk 2 above)
Comment (josdejong) [reply README.md:499]: Good point. It's hard to come up with a _real_ minimal example demonstrating referring to multiple signatures, but adding your example code is ok already I think.

File: test/construction.test.js (hunk 3 above)
Comment (josdejong) [reply test/construction.test.js:465]: Yes good point. I've been thinking shortly about this. You're right that this can will put the burden on the user, having to make sure there are no forward references. For a single function it's not a big problem I think, but it can get nasty when trying to merge two functions together. 

What I like about the current implementation though is that it is functional and very simple and predictable.

I'm not sure but it may turn out a relatively complex to do this auto-dependency-resolving neatly: you can get a chain of dependencies, you can get circular dependencies, etc. If OK with you I prefer to address that in a separate PR, and only as soon as the need arises. The current implementation is functional and we can improve on it in a later stage making it smarter and more convenient. At this moment I have the feeling that needing `referTo` is already a bit of an edge case, so I'm not sure if it will be a common case to have a typed-function with "lots" of signatures referring to each other. If the implementation for auto resolving these dependencies is relatively complex, I'm not sure if it is worth it. So I want to be a bit cautious in that regard spending time and increasing code complexity.

File: typed-function.js (hunk 4 above)
Comment (josdejong) [reply typed-function.js]: Very good point! Will fix that.

File: typed-function.js (hunk 5 above)
Comment (josdejong) [reply typed-function.js:1037]: You're right, will change the message.

File: typed-function.js (hunk 6 above)
Comment (josdejong) [reply typed-function.js:1103]: Hm, yes, a test for that is missing. You may need it when merging typed-functions. It is a bit of an edge case, you cannot produce an issue with this with the public API of creating and merging two functions directly. It can occur when you manually pick signatures from two existing typed-functions, and put them in one new typed-function.

(I had created a test for the same mechanism `resolvedSelf.referToSelf = fn.referToSelf` and just checked, the unit test fails when I remove that line of code).

I've now added a unit test for it, it is named _'should be able to use referTo when merging signatures from multiple typed-functions'._

**Comment by josdejong** [comment]
> So as an interim measure for compatibility errors, when incoming this is false-y, typed-function could replace it with a function that issues an error with a clear message about the change.

Ahh, that is an interesting idea. I'll do some playing around to see what works out without too much code/overhead 👍

**Comment by josdejong** [comment]
@gwhitney thanks for all your feedbacks, you found some important issues! 

1. I have addressed the small feedbacks 
2. I've refactored the solution to call `resolveReferences` after splitting the signatures. The function does still need a big refactor, it's quite a mess.
3. I've added a warning message when the old self-referencing via `this(...)` or `this.signatures` is used. I've thought about your idea of injecting a `this` which throws an exception, but I prefer a conservative solution that triggers an error at creation-time rather than run-time, and which has zero impact on the run-time performance. The solution I've implemented uses regex on the function bodies, so it will probably not catch 100% of the cases, but it works at creation-time and when you turn it off it has zero impact on both creation-time and run-time performance.

I'll let you know when I've done the refactoring of `resolveReferences`, would be great if you can have another look at the PR then.

File: typed-function.js (hunk 6 above)
Comment (gwhitney) [reply typed-function.js:1103]: Wow, that is quite an edge case, and frankly it feels a bit wrong that the outcome of `typed(tf1, tf2)` is different from `typed({...tf1.signatures, ..tf2.signatures})`. But anyhow, for the point in question, I am definitely not against keeping the .referTo property of the signature, so happy to leave this item be now.

File: test/construction.test.js (hunk 3 above)
Comment (gwhitney) [reply test/construction.test.js:465]: > At this moment I have the feeling that needing referTo is already a bit of an edge case, so I'm not sure if it will be a common case to have a typed-function with "lots" of signatures referring to each other.

I am confused; a good fraction of mathjs typed-functions use `this`, certainly dozens of them, and simplify() has numerous forward references, and derivative() has double references (implementation A refers to B refers to C) although those happen to all be arranged (so far as I can tell, it's a bit tricky) so all of the references are back references. I didn't keep searching for a double forward reference, but I would not be surprised if they are there. Adoption of v3 in mathjs is already going to be a big change (although well worth it), which means it would be nice if the changing of this --> referTo could be done as an editor macro, rather than having to worry about re-ordering the implementations of a typed-function as well.

As for the code to handle forward references, it's very simple: you just take the current resolution loop, and rather than erroring when unresolved, you set a flag, and also you set a flag when you perform a resolution. Then you just repeat the loop until the first time you perform no resolutions, and if the flag that there were unresolved references is set, you issue a "circular reference" error (you can use the item that was not resolved as the "true" value of the flag so that the error message can tell you where the circular reference occurred), and if there were no unresolved references, you're good to go.

I would be happy to add this to this PR if you would like; I really think we will save ourselves a significant amount of trouble with the mathjs adoption of v3 if we do.

Comment (gwhitney) [reply test/construction.test.js:465]: P.S. I think we mean different things by the word "functional," I don't see anything more or less functional about supporting forward references or not supporting them.

File: typed-function.js (hunk 4 above)
Comment (gwhitney) [reply typed-function.js]: P.S. if typed-function did allow forward references, one wouldn't have to worry about the relative order of sorting the signatures and resolving references, as is currently carefully commented in the code.

File: test/construction.test.js (hunk 3 above)
Comment (josdejong) [reply test/construction.test.js:465]: > I am confused; a good fraction of mathjs typed-functions use this, certainly dozens of them, and simplify() has numerous forward references

o wow, you're right. In my mind it wasn't used that much in mathjs but I just had a look at the code, there are 184 occurrences of `this(...)`. Thanks for correcting me. That will be quite some refactoring indeed. I hope we can use `referTo` a lot, that will improve runtime performance.

> As for the code to handle forward references, it's very simple: you just take the current resolution loop, and rather than erroring when unresolved, you set a flag, and also you set a flag when you perform a resolution.

It's probably not that complex as I first thought. Since `resolveReferences` still need a refactor, it can well be that it turn out to be the simplest solution is a recursive one which directly solves the forward references and circular dependencies.

> I would be happy to add this to this PR if you would like; I really think we will save ourselves a significant amount of trouble with the mathjs adoption of v3 if we do.

Thanks, a bit of help on this would be welcome! I'll make you a collaborator on `typed-function` too so you can directly work on this `feat/no_binding_self2` feature branch. I don't expect to start on this refactor today at least, but will let you know if I happen too after all.

> P.S. I think we mean different things by the word "functional," I don't see anything more or less functional about supporting forward references or not supporting them.

😂  that is a really nice one. I hadn't realized my confusing wording of _"it is functional and very simple"_, what I just meant there is _"it works and is very simple"_. Fits well though on Aprils Fools' Day 😅

File: typed-function.js (hunk 6 above)
Comment (josdejong) [reply typed-function.js:1103]: > and frankly it feels a bit wrong that the outcome of `typed(tf1, tf2)` is different from `typed({...tf1.signatures, ..tf2.signatures})`.

The reason is just how merging of two objects works: if both objects have the same key, the key of one of the two will be overwritten. So:

```js
const obj1 = { number: fn1 }
const obj2 = { number: fn2 }

// merging objects with the same keys will silently overwrite the first one with the second
const obj3 = { ...obj1, ...obj2 } // obj3 will be { number: fn2 }

const tf1 = typed(obj1)
const tf2 = typed(obj2)
const tf3 = typed(tf1, tf2) // will throw an exception about conflicting (duplicate) signatures
```

File: typed-function.js (hunk 4 above)
Comment (josdejong) [reply typed-function.js]: Yes, let's see if we can directly make that working! I'll close this conversation since the code works correctly now for this case, but the refactoring still has to be addressed.

File: test/construction.test.js (hunk 3 above)
Comment (gwhitney) [reply test/construction.test.js:465]: * Yes I think almost all instances of this in mathjs can become referTo -- just a small handful where they need to be referToSelf in their current structure.
* Thanks for the invitation, sometime over this weekend I will have a try at resolveReferences if you haven't.
* yes you did in fact inadvertently "April fool" me :)

Comment (gwhitney) [reply test/construction.test.js:465]: > Thanks, a bit of help on this would be welcome!

OK, I pushed a branch `feat/no_binding_self_refactor` to this repo. I didn't know if you'd want me to make a pull request to this branch or if you can just look at that branch and see if it looks good and switch to it if so, or what would be the most helpful, but I definitely thought you'd want to compare with the current state of this branch before deciding to go with it. In short, the refactor I propose also affects createTypedFunction -- basically I delay the introduction of conversions, since they don't do anything until it's time to construct the implementation, and keep the signature information just in a map until then, so that resolveReferences just has a signature map to deal with (not also an array of signatures). That seemed to me to simplify resolveReferences considerably. Anyhow, let me know if there's anything else I can do to assist with this, such as file a pull request to this branch or adjust the proposed approach to refactoring.

Comment (josdejong) [reply test/construction.test.js:465]: Thanks for keeping your refactoring in a separate branch, that indeed makes it easy for me to see what you did exactly. I'm really happy with this refactoring, it makes sense to move introducing conversions to the end!

I'll merge your refactor into the `feat/no_binding_self2` branch and do some final tuning (there is no transpiler in place so to keep IE11 happy we shouldn't use modern features, and I prefer `clearResolutions` and `resolveReferences` to not mutate the input arguments but create a new map)

Comment (josdejong) [reply test/construction.test.js:465]: Will now merge this PR, and address making things work on IE11 separately (there where also some other recent changes that are not IE11 compatible 😅 )

Comment (gwhitney) [reply test/construction.test.js:465]: Probably my fault. Sorry I don't have a clear idea in my head exactly what are IE11 transgressions... Is there a cheat sheet somewhere of "things not to do for IE11"?

Comment (josdejong) [reply test/construction.test.js:465]: It's not your fault, the library is just behind and has no guards for this and also no linting. I forgot this myself too. 

I've made `v3` working on IE11 again, see https://github.com/josdejong/typed-function/commit/67633e4058424bf79a2bc0cd2ce20e184a033193. It's kind of stupid to go that way. On the other hand it makes me realize how much improvement JavaScript latest years 😁 .

To test I just try to open the file test/browser.html in IE 11 (via BrowserStack) and see if it works.
